content/blog/taksim-modes-makam.md

```diff
@@ -0,0 +1,664 @@
+
+To encode this rule properly, we need to make sure
+that the type `T` of `type_check E T` is treated as an input.
+The key difference is that
+given an unknown type like `T` as input, we should not unify it with the
+`arrow A1 A2` term; instead, we should check that `T` *already*
+looks like `arrow A1 A2`. So we should perform *pattern matching*
+rather than *unification* with `arrow A1 A2`. Fortunately,
+Makam provides us with a `pattern_match` implementation that behaves
+as expected:
+
+```makam
+unify (arrow A1 A2) T ?
+pattern_match (arrow A1 A2) T ?
+pattern_match (arrow A1 A2) (arrow tunit tunit) ?
+pattern_match (arrow A1 A2) (arrow C D) ?
+```
+
+```makam-hidden
+>> unify (arrow A1 A2) T ?
+>> Yes:
+>> T := arrow A1 A2, A1 := A1, A2 := A2.
+
+>> pattern_match (arrow A1 A2) T ?
+>> Impossible.
+
+>> pattern_match (arrow A1 A2) (arrow tunit tunit) ?
+>> Yes:
+>> A1 := tunit, A2 := tunit.
+
+>> pattern_match (arrow A1 A2) (arrow C D) ?
+>> Yes:
+>> A1 := C, A2 := D, C := C, D := D.
+```
+
+(The first argument to `pattern_match` is the pattern, and the second is the scrutinee).
+
+So encoding the arrow introduction rule accurately is simple -- we
+can just use `pattern_match` instead of unifying directly:
+
+```makam
+type_check (lambda E) T :-
+  pattern_match (arrow A1 A2) T,
+  (x:expr -> type_of_var x A1 -> type_check (E x) A2).
+```
+
+To be entirely accurate, we should use pattern matching for the
+expression argument, since we expect that to be an input too. But
+we'll keep that as-is for now and revisit that later.
+
+Transcribing the rest of the rules, we proceed similarly for the
+checking rules, and there is not much we need to change for synthesis rules.
+
+```makam
+type_synth X A :-
+  type_of_var X A.
+
+type_check E B :-
+  type_synth E A, unify A B.
+
+type_synth (annot E A) A :-
+  type_check E A.
+
+type_check eunit T :-
+  pattern_match tunit T.
+
+type_synth (app E1 E2) B :-
+  type_synth E1 (arrow A B),
+  type_check E1 A.
+```
+
+Let's try a few examples to see what we did:
+
+```makam
+type_check (lambda (fun x => x)) (arrow tunit tunit) ?
+type_check (lambda (fun x => x)) T ?
+type_check (lambda (fun x => x)) (arrow tunit X) ?
+type_check (lambda (fun x => x)) (arrow tunit (arrow tunit tunit)) ?
+```
+
+These all behave as expected.
+
+```makam-hidden
+>> type_check (lambda (fun x => x)) (arrow tunit tunit) ?
+>> Yes.
+
+>> type_check (lambda (fun x => x)) T ?
+>> Impossible.
+
+>> type_check (lambda (fun x => x)) (arrow tunit X) ?
+>> Yes:
+>> X := tunit.
+
+>> type_check (lambda (fun x => x)) (arrow (arrow tunit tunit) X) ?
+>> Yes:
+>> X := arrow tunit tunit.
+
+>> type_check (lambda (fun x => x)) (arrow tunit (arrow tunit tunit)) ?
+>> Impossible.
+```
+
+```makam-hidden
+%end.
+```
+
+So the recipe for enforcing a proper mode is pretty simple: for input arguments, we should use pattern matching rather than unification.
+Is there some way that we can extend Makam to support this recipe?
+
+## modal makam
```

Comment:
Ah yes, this is confusing, thanks for pointing it out. I'm not aware and don't mean to suggest that there's much of a relationship between logic programming modes and modal logic etc. (I think the naming of the latter derives from the notion of a modality instead of that of modes).

This was mostly a music-inspired word play (the use of musical modes is one of the similarities between modal jazz and makam-based composition, and I find that the Ibrahim Maalouf album leans on that) but it's not worth the confusion.